larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -58,12 +68,84 @@ DenseElementsAttr Bitpack(PatternRewriter& builder, Attribute x) {
 
 #include "larq_compute_engine/mlir/transforms/generated_optimize.inc"
 
+struct SetBconvReadWriteBitpacked : public OpRewritePattern<TF::LceBconv2dOp> {
+  using OpRewritePattern<TF::LceBconv2dOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(TF::LceBconv2dOp bconv_op,
+                                     PatternRewriter& rewriter) const override {
+    Value bconv_input = bconv_op.input();
+    if (!bconv_input.hasOneUse()) return matchFailure();
+
+    auto bconv_input_op =
+        dyn_cast_or_null<TF::LceBconv2dOp>(bconv_input.getDefiningOp());
+    if (!bconv_input_op) return matchFailure();
+
+    if (bconv_input_op.write_bitpacked_output() &&
+        bconv_op.read_bitpacked_input())
+      return matchFailure();
+
+    if (bconv_input_op.padding() == "SAME" && bconv_input_op.pad_values() == 0)
+      return matchFailure();
+
+    const auto inner_tensor_shape =
+        bconv_input_op.getType().cast<ShapedType>().getShape();
+    if (inner_tensor_shape.size() != 4) return matchFailure();
+
+    // We use 32-bit bitpacking.
+    constexpr int bitwidth = 32;
+
+    const auto channels = inner_tensor_shape[3];
+    const auto packed_channels = (channels + bitwidth - 1) / bitwidth;
+
+    RankedTensorType inner_tensor_type =
+        RankedTensorType::get({inner_tensor_shape[0], inner_tensor_shape[1],
+                               inner_tensor_shape[2], packed_channels},
+                              rewriter.getIntegerType(bitwidth));
+
+    rewriter.replaceOpWithNewOp<TF::LceBconv2dOp>(
```

Comment:
Do you use `replaceOpWithNewOp` since you need to replace two ops at once and `replaceOp` won't work?

Reply:
Ah I wasn't actually aware of `replaceOp` because I copied the boilerplate for this pass from a TF example. I can't find a clear definition of `replaceOp` though and it seems to only accept two arguments; do you know how it's meant to be used?

Reply:
:+1: I'll check if there is a qay to simplify this.

Reply:
I wonder if the two `replaceOpWithNewOp` statements can just be replaced by:
```cpp
bconv_input_op.setAttr("write_bitpacked_output", rewriter.getBoolAttr(true))
bconv_op.setAttr("read_bitpacked_input", rewriter.getBoolAttr(true))
```

Reply:
Now that we're removing the two boolean flags entirely and relying on the types, I suspect that the `replaceOpWithNewOp` will be required now as I doubt you can change the output type? I could be wrong though.